ops/pebble.py

```diff
@@ -1466,45 +1467,39 @@ def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.Bin
         if not boundary:
             raise ProtocolError('invalid boundary {!r}'.format(boundary))
 
-        # We have to manually write the Content-Type with boundary, because
-        # email.parser expects the entire multipart message with headers.
-        parser = email.parser.BytesFeedParser()
-        parser.feed(b'Content-Type: multipart/form-data; boundary='
-                    + boundary.encode('utf-8') + b'\r\n\r\n')
+        parser = MultipartFileParser(boundary)
 
-        # Then read the rest of the response and feed it to the parser.
         while True:
             chunk = response.read(self._chunk_size)
             if not chunk:
                 break
             parser.feed(chunk)
-        message = parser.close()
 
-        # Walk over the multipart parts and read content and metadata.
-        resp = None
-        content = None
-        for part in message.walk():
-            name = part.get_param('name', header='Content-Disposition')
-            if name == 'response':
-                resp = _json_loads(part.get_payload())
-            elif name == 'files':
-                filename = part.get_filename()
-                if filename != path:
-                    raise ProtocolError('path not expected: {}'.format(filename))
-                # decode=True, ironically, avoids decoding bytes to str
-                content = part.get_payload(decode=True)
+        filenames = parser.filenames()
+        if filenames:
+            filename = filenames[0]
+            if filename != path:
+                raise ProtocolError('path not expected: {}'.format(filename))
 
+        resp = parser.get_response()
         if resp is None:
             raise ProtocolError('no "response" field in multipart body')
         self._raise_on_path_error(resp, path)
 
-        if content is None:
+        if len(filenames) < 1:
             raise ProtocolError('no file content in multipart response')
-        if encoding is not None:
-            reader = io.StringIO(content.decode(encoding))
-        else:
-            reader = io.BytesIO(content)
-        return reader
+        if len(filenames) > 1:
+            raise NotImplementedError('multiple file responses not yet supported')
+
+        f = parser.get_file(path, encoding)
+
+        # the opened file remains usable after removing/unlinking until it is
+        # closed.  This prevents callers from interacting with it on disk.
+        # But e.g. windows doesn't allow removing opened files like this.
+        if os.name != 'posix' or sys.platform == 'cygwin':
+            return tempfile._TemporaryFileWrapper(f, f.name, delete=True)
```

Comment:
Hmm, I don't love this, though I don't know of a better way offhand.

1) Using `os.name` and `sys.platform` is a bit of a code smell - is there any way to use feature detection instead?
2) `tempfile._TemporaryFileWrapper` is an undocumented, "private" class, so it doesn't seem wise to use it. Then again, I can see why you have, as it's a fair bit of gnarly code. And it's been there for years, so unlikely to go away anytime soon ... as such, I guess depending on it is reasonable.

Reply:
I don't love it too much either.  I thought about this quite a bit - and while it does seem a bit hacky - this is in fact exactly how the tempfile lib discovers how to treat temp file opening/deletion differently for windows. I basically pasted this check from the stdlib.

Reply:
Also - on the undoc'd stdlib function `_TemporaryFileWrapper` - I did want to make sure that that class/interface was relatively dependable - and indeed the API as far as we are using has been stable since well back into python2.  I'm definitely open to alternative suggestions, but this seemed like a reasonable approach.

Reply:
Fair -- thanks for the due diligence!